pkg/nativeimgutil/nativeimgutil.go

```diff
@@ -0,0 +1,156 @@
+// Package nativeimgutil provides image utilities that do not depend on `qemu-img` binary.
+package nativeimgutil
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/containerd/continuity/fs"
+	"github.com/lima-vm/go-qcow2reader"
+	"github.com/lima-vm/go-qcow2reader/image/qcow2"
+	"github.com/lima-vm/go-qcow2reader/image/raw"
+	"github.com/lima-vm/lima/pkg/osutil"
+	"github.com/sirupsen/logrus"
+)
+
+// ConvertToRaw converts a source disk into a raw disk.
+// source and dest may be same.
+// ConvertToRaw is a NOP if source == dest, and no resizing is needed.
+func ConvertToRaw(source, dest string, size *int64, allowSourceWithBackingFile bool) error {
+	srcF, err := os.Open(source)
+	if err != nil {
+		return err
+	}
+	defer srcF.Close()
+	srcImg, err := qcow2reader.Open(srcF)
+	if err != nil {
+		return fmt.Errorf("failed to detect the format of %q: %w", source, err)
+	}
+	if size != nil && *size < srcImg.Size() {
+		return fmt.Errorf("specified size %d is smaller than the original image size (%d) of %q", *size, srcImg.Size(), source)
+	}
+	switch t := srcImg.Type(); t {
+	case raw.Type:
+		if err = srcF.Close(); err != nil {
+			return err
+		}
+		return convertRawToRaw(source, dest, size)
+	case qcow2.Type:
+		if !allowSourceWithBackingFile {
+			q, ok := srcImg.(*qcow2.Qcow2)
+			if !ok {
+				return fmt.Errorf("unexpected qcow2 image %T", srcImg)
+			}
+			if q.BackingFile != "" {
+				return fmt.Errorf("qcow2 image %q has an unexpected backing file: %q", source, q.BackingFile)
+			}
+		}
+	default:
+		logrus.Warnf("image %q has an unexpected format: %q", source, t)
+	}
+	if err = srcImg.Readable(); err != nil {
+		return fmt.Errorf("image %q is not readable: %w", source, err)
+	}
+
+	// Create a tmp file because source and dest can be same.
+	destTmpF, err := os.CreateTemp(filepath.Dir(dest), filepath.Base(dest)+".lima-*.tmp")
+	if err != nil {
+		return err
+	}
+	destTmp := destTmpF.Name()
+	defer os.RemoveAll(destTmp)
+	defer destTmpF.Close()
+
+	// Copy
+	srcImgR := io.NewSectionReader(srcImg, 0, srcImg.Size())
+	const bufSize = 1024 * 1024
+	if copied, err := copySparse(destTmpF, srcImgR, bufSize); err != nil {
+		return fmt.Errorf("failed to call copySparse(), bufSize=%d, copied=%d: %w", bufSize, copied, err)
+	}
+
+	// Resize
+	if size != nil {
+		if err = MakeSparse(destTmpF, *size); err != nil {
+			return err
+		}
+	}
+	if err = destTmpF.Close(); err != nil {
```

Comment:
In case of vm starting stopped during conversion of disk, the temp files are not cleared

Reply:
The line 65 has `os.RemoveAll(destTmp)`

Reply:
Ya i noticed but on testing this usecase, i could see stale temp diffdisk files.

Steps i did,
- limactl  start template://experimental/vz
- ctrl+c when we get that converting log
- Noticed the diffdisk-*.temp in directory

Reply:
`limactl stop` now remove `*.tmp`